backend/linkbox/linkbox.go

```diff
@@ -0,0 +1,946 @@
+	fs.Infof(o, "Open - {%s}", o.Remote())
+
+	var res *http.Response
+	downloadURL := o.fullURL
+	if downloadURL == "" {
+		_, name := splitDirAndName(o.Remote())
+		newObject, err := getObject(ctx, o.fs, name, o.pid, o.fs.opt.Token)
+		if err != nil {
+			return nil, err
+		}
+		if newObject == (Entity{}) {
+			fs.Debugf(o.fs, "Open entity is empty: name = {%s}", name)
+			return nil, fs.ErrorObjectNotFound
+		}
+
+		downloadURL = newObject.URL
+	}
+
+	opts := &rest.Opts{
+		Method:  "GET",
+		RootURL: downloadURL,
+		Options: options,
+	}
+
+	err := o.fs.pacer.Call(func() (bool, error) {
+		var err error
+		res, err = o.fs.srv.Call(ctx, opts)
+		return o.fs.shouldRetry(ctx, err)
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Open failed: %w", err)
+	}
+
+	return res.Body, nil
+}
+
+// Update in to the object with the modTime given of the given size
+//
+// When called from outside an Fs by rclone, src.Size() will always be >= 0.
+// But for unknown-sized objects (indicated by src.Size() == -1), Upload should either
+// return an error or update the object properly (rather than e.g. calling panic).
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	fs.Infof(o, "Update: Remote = {%s}", o.Remote())
+
+	if src.Size() == 0 {
+		return fs.ErrorCantUploadEmptyFiles
+	}
+
+	remote := o.Remote()
+	tmpObject, err := o.fs.NewObject(ctx, remote)
+
+	if err == nil {
+		fs.Debugf(o.fs, "Update: removing old file")
+		_ = tmpObject.Remove(ctx)
+	}
+
+	h := md5.New()
+	first10m := io.LimitReader(in, 10_485_760)
+	first10mBytes, err := io.ReadAll(first10m)
+	if err != nil {
+		fs.Debugf(o.fs, "Update err in reading file: %v", err)
+		return err
+	}
+
+	if _, err := io.Copy(h, bytes.NewReader(first10mBytes)); err != nil {
+		fs.Debugf(o.fs, "Update err in copying %v", err)
+		return err
+	}
+
+	// get upload authorization (step 1)
+	opts := &rest.Opts{
+		Method:  "GET",
+		RootURL: linkboxAPIURL,
+		Path:    "get_upload_url",
+		Parameters: url.Values{
+			"token":           {o.fs.opt.Token},
+			"fileMd5ofPre10m": {fmt.Sprintf("%x", h.Sum(nil))},
+			"fileSize":        {strconv.FormatInt(src.Size(), 10)},
+		},
+	}
+
+	getFistStepResult := getUploadUrlResponse{}
+	err = getUnmarshaledResponse(ctx, o.fs, opts, &getFistStepResult)
+	if err != nil {
+		fs.Debugf(o.fs, "Update err in unparshaling response: %v", err)
+		return err
+	}
+
+	switch getFistStepResult.Status {
+	case 1:
+		// upload file using link from first step
+		var res *http.Response
+
+		file := io.MultiReader(bytes.NewReader(first10mBytes), in)
+
+		opts := &rest.Opts{
+			Method:  "PUT",
+			RootURL: getFistStepResult.Data.SignUrl,
+			Body:    file,
+		}
+
+		err = o.fs.pacer.CallNoRetry(func() (bool, error) {
+			res, err = o.fs.srv.Call(ctx, opts)
+			return o.fs.shouldRetry(ctx, err)
+		})
+
+		if err != nil {
+			fs.Debugf(o.fs, "Update err in uploading file: %v", err)
+			return err
```

Comment:
Instead of debugging the error message, decorate it, so

    return fmt.Errorf("update err in uploading file: %v", err)

Then these more helpful errors will appear in users logs.

There are more of these but I won't comment on any more.

Reply:
I used %w and it still causes tests failure. I didn't look very precisely, but seems it is because errors are compared by bites it tests.

Reply:
Some errors need to be returned exactly and some don't. Just make the tests pass and all will be well :-)